src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildLogsExtension.java

```diff
@@ -72,22 +85,140 @@ public BuildLogResponse getBuildLog(
         final long skipF = skip;
         return getBuildByNumberOrLast(jobFullName, buildNumber)
                 .map(build -> {
-                    try (BufferedReader reader = new BufferedReader(build.getLogReader())) {
-                        List<String> allLines = reader.lines().toList();
-                        long actualOffset = skipF;
-                        if (skipF < 0) actualOffset = Math.max(0, allLines.size() + skipF);
-                        int endIndex = (int) Math.min(actualOffset + limitF, allLines.size());
-                        var lines = allLines.subList((int) actualOffset, endIndex);
-                        boolean hasMoreContent = endIndex < allLines.size();
-                        return new BuildLogResponse(hasMoreContent, lines);
-
-                    } catch (IOException e) {
+                    try {
+                        return getLogLines(build, skipF, limitF);
+                    } catch (Exception e) {
                         log.error("Error reading log for job {} build {}", jobFullName, buildNumber, e);
                         return null;
                     }
                 })
                 .orElse(null);
     }
 
+    private BuildLogResponse getLogLines(Run<?, ?> run, long skip, int limit) throws Exception {
+
+        int maxLimit = SystemProperties.getInteger(BuildLogsExtension.class.getName() + ".limit.max", 10000);
+
+        limit = Math.min(Math.abs(limit), maxLimit);
+
+        if (skip < 0) {
+            // TODO can get gz file?
+            File logFile = run.getLogFile();
```

Comment:
that's a bit of a problem though. as this generating a lot warning in the logs when accessing logs of a pipeline because of https://github.com/jenkinsci/workflow-job-plugin/blame/b07d2bbfadb9d1074bd96184d73eb30cb203223d/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L1220